File: app/controllers/LittleGolemParser.rb
#!/usr/bin/ruby
require 'TwixtBoard'
require 'Peg'
require 'Move'

class LittleGolemParser
  def initialize(game)
    @game = game
  end

  def getPlayer1
    p = @game.scan(/PB\[[^\]]*\]/)[0]
    return p[3..(p.length - 2)]
  end

  def getPlayer2
    p = @game.scan(/PW\[[^\]]*\]/)[0]
    return p[3..(p.length - 2)]
  end
  
  def getTournament
    p = @game.scan(/EV\[[^\]]*\]/)[0]
    tourney = p[3..(p.length - 2)]
    return (tourney == 'null')? '': tourney
  end

  def getMoves
    moves = @game.scan(/[br]\[[^\]]*\]/)
  end

  def getWinner
    p = @game.scan(/[br]\[resign\]/)[0]
    if p == nil
      return nil
    else
      return " rb".index(p[0])
    end
  end

  def forEachMove
    letters=" abcdefghijklmnopqrstuvwx"
    moves = getMoves
    swapped = false

    # swap rule
    if moves.length > 1 and moves[1] == "r[swap]"
       move = moves[0]
       x = letters.index(move[2]) - 1
       y = letters.index(move[3]) - 1
       yield Move.new(x, y, 1)  # place white peg
       yield SwapMove.new(2) # player 2 swaps
       moves.delete_at(1)  # remove the swap move
       moves.delete_at(0)  # remove the first move
       swapped = true
    end

    player = 1
    # the rest of the moves
    for move in moves
      if swapped
        player = " rb".index(move[0])
      else
        player = " br".index(move[0])
      end
      
      if move.length == 5 or move[4..9] == "|draw]"
        x = letters.index(move[2]) - 1
        y = letters.index(move[3]) - 1
        if swapped
          yield Move.new(y, x, player)
        else
          yield Move.new(x, y, player)
        end
      elsif move[1..8] == "[resign]"
        yield ResignMove.new(player)
      elsif move[1..6] == "[draw]"
        yield DrawMove.new(player)
      end
    end

    if @forfeit
      yield ForfeitMove.new(3 - player)
    end
  end

  def getTwixtBoard
    board = TwixtBoard.new(24, 24)

    board.setPlayer(1, getPlayer1)
    board.setPlayer(2, getPlayer2)

    forEachMove { |move|
      board.play(move)
    }
    
    board
  end
  
  def isGameOver
    board = getTwixtBoard
    board.isDraw | board.hasWonColor(:white) | board.hasWonColor(:black)
  end

  def isGameLostOnMoves
    board = getTwixtBoard
    board.hasConnectedColor(:white) | board.hasConnectedColor(:black)
  end

  # much simpler parser -- keeps the swap move in there, keeps LG conventions of swapped boards.
  def forEachLittleGolemMove
    letters=" abcdefghijklmnopqrstuvwx"
    moves = getMoves

    player = 1
    for move in moves
      player = " br".index(move[0])

      if move.length == 5 or move[4..9] == "|draw]"
        x = letters.index(move[2])
        y = letters.index(move[3])
        yield Move.new(x, y, player)
      elsif move[1..8] == "[resign]"
        yield ResignMove.new(player)
      elsif move[1..6] == "[draw]"
        yield DrawMove.new(player)
      elsif move == "r[swap]"
        yield SwapMove.new(player)
      end
    end
    if (@forfeit)
      yield ForfeitMove.new(3 - player)
    elsif (isGameLostOnMoves)
      yield LostMove.new(3 - player)
    end
  end

  def getMovesList
    moves = []
    forEachLittleGolemMove { |move| moves << move }
    moves
  end
  
  def forfeit
    @forfeit = true
  end
  
  def getResultChar
    board = getTwixtBoard
    
    (board.isResignation)? 'R':
    (board.hasConnectedColor(:white) | board.hasConnectedColor(:black))? 'L':
    (board.isDraw)? 'D':
    (@forfeit)? 'F':
    '?'
  end
  
  
  def forEachMoveForJtwixt
    letters=" abcdefghijklmnopqrstuvwx"
    moves = getMoves

    # swap rule
    if moves.length > 1 and moves[1] == "r[swap]"
      move = moves[0]
      x = letters.index(move[3])
      y = letters.index(move[2])
      yield x, y, 1       # place white peg first so Jtwixt doesn't get confused
      yield x, y, 2       # then put black peg right on top of it.
      moves.delete_at(1)  # remove the swap move
      moves.delete_at(0)  # remove the first move
    end

    # the rest of the moves
    for move in moves
      if move.length > 4 and (move[4..4] == "]" or move[4..4] == "|") 
        player = " br".index(move[0])
        x = letters.index(move[2])
        y = letters.index(move[3])
        yield x, y, player
      end
    end
  end
end

File: lib/domain/LittleGolemParser.js
const { TwixtBoard } = require('./TwixtBoard');
const { Move, SwapMove, ResignMove, DrawMove, ForfeitMove, LostMove } = require('../../public/javascripts/move');

class LittleGolemParser {
  constructor(game) {
    this._game = game;
    this._forfeit = false;
  }

  getPlayer1() {
    const m = this._game.match(/PB\[[^\]]*\]/);
    if (!m) return null;
    const p = m[0];
    return p.slice(3, p.length - 1);
  }

  getPlayer2() {
    const m = this._game.match(/PW\[[^\]]*\]/);
    if (!m) return null;
    const p = m[0];
    return p.slice(3, p.length - 1);
  }

  getBoardSize() {
    const m = this._game.match(/SZ\[[^\]]*\]/);
    const p = m[0];
    return parseInt(p.slice(3, p.length - 1), 10);
  }

  _getMoveFromLetter(letter) {
    // zero based index: 'a' (97) --> 0, 'A' (65) --> 26
    return (letter >= 'a' && letter <= 'z')?
      letter.charCodeAt(0) - 97:
      letter.charCodeAt(0) - 39;
  }

  getTournament() {
    const m = this._game.match(/EV\[[^\]]*\]/);
    if (!m) return '';
    const p = m[0];
    const tourney = p.slice(3, p.length - 1);
    return tourney === 'null' ? '' : tourney;
  }

  getMoves() {
    return this._game.match(/[br]\[[^\]]*\]/g) || [];
  }

  // Iterates moves, calling cb(move) for each Move object.
  // Mirrors forEachMove in Ruby (handles swap rule, player attribution).
  forEachMove(cb) {
    let moves = this.getMoves();
    let swapped = false;

    // swap rule
    if (moves.length > 1 && moves[1] === 'r[swap]') {
      const m = moves[0];
      const x = this._getMoveFromLetter(m[2]);
      const y = this._getMoveFromLetter(m[3]);
      cb(new Move(x, y, 1));       // place white peg
      cb(new SwapMove(2));          // player 2 swaps
      moves.splice(0, 2);           // remove first two
      swapped = true;
    }

    let player = 2;
    for (const move of moves) {
      if (swapped) {
        player = ' rb'.indexOf(move[0]);
      } else {
        player = ' br'.indexOf(move[0]);
      }

      if (move.length === 5 || move.slice(4, 10) === '|draw]') {
        const x = this._getMoveFromLetter(move[2]);
        const y = this._getMoveFromLetter(move[3]);
        if (swapped) {
          cb(new Move(y, x, player));
        } else {
          cb(new Move(x, y, player));
        }
      } else if (move.slice(1, 9) === '[resign]') {
        cb(new ResignMove(player));
      } else if (move.slice(1, 7) === '[draw]') {
        cb(new DrawMove(player));
      }
    }

    if (this._forfeit) {
      cb(new ForfeitMove(3 - player));
    }
  }

  getTwixtBoard() {
    const size = this.getBoardSize();
    const board = new TwixtBoard(size, size);
    board.setPlayer(1, this.getPlayer1());
    board.setPlayer(2, this.getPlayer2());
    this.forEachMove(move => board.play(move));
    return board;
  }

  isGameOver() {
    const board = this.getTwixtBoard();
    return board.isDraw() || board.hasWonColor('white') || board.hasWonColor('black');
  }

  isGameLostOnMoves() {
    const board = this.getTwixtBoard();
    return board.hasConnectedColor('white') || board.hasConnectedColor('black');
  }

  // Simpler parser — keeps swap move and LG conventions of swapped boards.
  forEachLittleGolemMove(cb) {
    const moves = this.getMoves();
    let player = 1;

    for (const move of moves) {
      player = ' br'.indexOf(move[0]);

      if (move.length === 5 || move.slice(4, 10) === '|draw]') {
        const x = this._getMoveFromLetter(move[2]) + 1;
        const y = this._getMoveFromLetter(move[3]) + 1;
        cb(new Move(x, y, player));
      } else if (move.slice(1, 9) === '[resign]') {
        cb(new ResignMove(player));
      } else if (move.slice(1, 7) === '[draw]') {
        cb(new DrawMove(player));
      } else if (move === 'r[swap]') {
        cb(new SwapMove(player));
      }
    }

    if (this._forfeit) {
      cb(new ForfeitMove(3 - player));
    } else if (this.isGameLostOnMoves()) {
      cb(new LostMove(3 - player));
    }
  }

  getMovesList() {
    const moves = [];
    this.forEachLittleGolemMove(m => moves.push(m));
    return moves;
  }

  // Convert a zero-based coordinate index back to its letter representation.
  // 0-25 -> 'a'-'z', 26-51 -> 'A'-'Z'
  _letterFromIndex(index) {
    if (index >= 0 && index <= 25) {
      return String.fromCharCode(97 + index);
    }
    return String.fromCharCode(39 + index);
  }

  // Normalize a coordinate index to the upper-left quadrant.
  // On a board of size N, index i becomes min(i, N-1-i).
  _normalizeCoordinate(index, boardSize) {
    return Math.min(index, boardSize - 1 - index);
  }

  // Return the first move normalized to the upper-left quadrant as a 2-char string.
  // Returns null if there is no first move (e.g., empty game).
  getNormalizedFirstMove() {
    const moves = this.getMoves();
    if (!moves.length) return null;

    const first = moves[0];

    // Only coordinate moves (e.g. b[ab] or b[ab|draw]) have a normalized first move.
    // Non-coordinate moves like b[resign] return null.
    if (first.length !== 5 && first.slice(4, 10) !== '|draw]') return null;

    const boardSize = this.getBoardSize();

    const xNorm = this._normalizeCoordinate(this._getMoveFromLetter(first[2]), boardSize);
    const yNorm = this._normalizeCoordinate(this._getMoveFromLetter(first[3]), boardSize);

    return this._letterFromIndex(xNorm) + this._letterFromIndex(yNorm);
  }

  // Return whether the game used a swap (second SGF move is r[swap]).
  getSwapped() {
    const moves = this.getMoves();
    return moves.length > 1 && moves[1] === 'r[swap]';
  }

  // Called externally when we've determined the game is a forfeit via HTML scrape.
  forfeit() {
    this._forfeit = true;
  }

  getResultChar() {
    const board = this.getTwixtBoard();
    if (board.isResignation()) return 'R';
    if (board.hasConnectedColor('white') || board.hasConnectedColor('black')) return 'L';
    if (board.isDraw()) return 'D';
    if (this._forfeit) return 'F';
    return '?';
  }

  // Used by jtwixt route to generate JTwixt file data.
  forEachMoveForJtwixt(cb) {
    let moves = this.getMoves();

    // swap rule
    if (moves.length > 1 && moves[1] === 'r[swap]') {
      const m = moves[0];
      const x = this._getMoveFromLetter(m[3]) + 1;
      const y = this._getMoveFromLetter(m[2]) + 1;
      cb(x, y, 1);   // place white peg first
      cb(x, y, 2);   // then black on top
      moves.splice(0, 2);
    }

    for (const move of moves) {
      if (move.length > 4 && (move[4] === ']' || move[4] === '|')) {
        const player = ' br'.indexOf(move[0]);
        const x = this._getMoveFromLetter(move[2]) + 1;
        const y = this._getMoveFromLetter(move[3]) + 1;
        cb(x, y, player);
      }
    }
  }
}

module.exports = { LittleGolemParser };
